Added RSpec Tests

s0kil opened the pull request:
- Added specs for all Enumerable methods we previously created.
- Added README instructions to run the tests.

s0kil committed:
Initial Testing Scafold

cod3smith committed:
Add test for remaining enumerable methods.

cod3smith committed:
Add tests.yml.

cod3smith committed:
Add trigger event to tests.yml.

cod3smith committed:
Add rspec to Gemfile.

cod3smith committed:
Add rspec to Gemfile.

cod3smith committed:
Add rspec to Gemfile.

cod3smith committed:
Add rspec to Gemfile.

s0kil committed:
Added Gem Comment

davidmrtz-dev reviewed (changes requested):
## Hi Team .
### Very good job so far.

However, there are some issues that need to be addressed before move on to the next project.  👨‍💻  ⬇️ 

****

- [x] Readme file.
Please update the readme file in order to add the instructions to run the tests suite (RSpec), you can add the command "RSpec" after "bundle install" in order to let the user run the test suite and also see the output result through the terminal.

****

- [x] Pull Request has a meaningful title and a brief description ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_prs.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
Your Pull Request is good but we can improve it.
Also, you can add a shortlist that contains valuable information about what you did.
****

### **Please make the changes required and submit for re-review once you are ready!** 

****

### Status: Changes required. ♻️  
****

Happy coding!  💻  
Have a great day 🙂

If you need more clarifications kindly reach me on slack.
**[TSE:  David Elí]** 
Slack: @ David Eli Martinez Garcia

s0kil committed:
Add Running Test Instruction

santiago-dev96 commented:
# Pull Request under review by @santiago-rodrig (TSE)

santiago-dev96 commented on README.md:
```diff
@@ -3,25 +3,32 @@
 Create #my_each
-Create #my_each_with_index 
-Create #my_select 
-Create #my_all? 
+Create #my_each_with_index
+Create #my_select
+Create #my_all?
 Create #my_any?
 Create #my_none?
 Create #my_count
 Create #my_map
 Create #my_inject
```
Replace all this with a listing of the imitated methods, like follows.

- `#each` -> `#my_each`
- `#inject` -> `#my_inject`
...

santiago-dev96 commented on README.md:
```diff
@@ -3,25 +3,32 @@
 The goal of this project is to add new methods to the existing Enumerable module, a method that is identical to the next ones, but (obviously) does not use the original ones.
```
```suggestion
The goal of this project is to make a working copy of some of the most popular methods from the `Enumerable` module of the Ruby programming language.

The list that follows shows what methods were copied.
```

santiago-dev96 commented on README.md:
```diff
@@ -3,25 +3,32 @@
 ## Run Debugger
 
-`rdebug-ide enumerables.rb`
+- `rdebug-ide enumerables.rb`
```
```suggestion
- `rdebug-ide enumerable.rb`
```

santiago-dev96 commented on README.md:
```diff
@@ -3,25 +3,32 @@
 ## Built With
 
 - Ruby
 - Rubocop
-- Test-RSpec
+- RSpec
```
Specify the versions, here is an example.

- Ruby (2.6.6)
- Rubocop (0.81)

santiago-dev96 commented on README.md:
```diff
@@ -38,7 +45,7 @@ Create #my_inject
-1. Open the terminal comand line window
+1. Open the terminal command line window
 2. `git clone https://github.com/epinczinger/ruby_enumerables.git`
 3. Execute `ruby enumerables.rb`
```
For this project, you are not creating an executable file, it doesn't make sense to issue `ruby enumerables.rb`, because this is a module, it doesn't interact with the user in any form.

```suggestion
3. run the tests to make sure all the methods do what they are supposed to do
```

santiago-dev96 reviewed (changes requested):
# Changes Required :wrench:

Your tests look good! Yet your codebase is messy.

1. delete the `run.rb` file and the `enumerable_spec.rb` at the root of your project

2. Fix your `README.md` file

### Optional

About the `run.rb` file, if you want to run tasks (like npm scripts) you should use [rake](https://www.stuartellis.name/articles/rake/), it is part of the standard library and is more professional.

## What to do now?

Implement the required changes and submit another CRR.

You can find me in slack as **srodrig**, happy coding! :wave:

s0kil commented:
@santiago-rodrig What does this PR have to do with the codebase? This is a "testing with RSpec" project.

santiago-dev96 commented:
> @santiago-rodrig What does this PR have to do with the codebase? This is a "testing with RSpec" project.

Yes @s0kil, but I have to make sure before allowing you to merge into develop that you are being professional with your code and that you are following best practices. I don't see any reason why you would like to have files that do nothing in your project. Also, every PR has to do with the codebase, in the end, the merged code gets into the codebase, think like if you were part of a team of developers working for a company, no one wants a cluttered codebase.

BrittanyBlake commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,202 @@
+require_relative('../enumerable')
+
+describe ::Enumerable do
+  describe "#my_each" do
+    it "Loops An Enumerable" do
+      array = []
+      %w[a b c].my_each { |item| array << item }
+      expect(array).to(eq(%w[a b c]))
+    end
+
+    it "Loops An Range" do
+      array = []
+      (1...4).my_each { |item| array << item }
+      expect(array).to(eq([1, 2, 3]))
+    end
+
+    it "Loops An Hash" do
+      array = []
+      { "Jane Doe" => 10, "Jim Doe" => 6 }.my_each { |item| array << item[0] }
+      expect(array).to(eq(["Jane Doe", "Jim Doe"]))
+    end
+  end
```
Please add a negative scenario for this spec which can be something like:  `it 'should not return an element that is not in the array'`

BrittanyBlake replied:
Please add a negative scenario for this spec which can be something like:  `it 'should not return an element that is not in the array'`

BrittanyBlake commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,202 @@
+    end
+  end
+
+  describe "#my_each_with_index" do
+    it "Loops An Hash With Index" do
+      hash = {}
+      [9, 8, 7, 6, 5, 4, 3, 2, 1].my_each_with_index do |item, index|
+        hash[index] = item
+      end
+
+      expect(hash).to(eq({ 0 => 9, 1 => 8, 2 => 7, 3 => 6, 4 => 5, 5 => 4, 6 => 3, 7 => 2, 8 => 1 }))
+    end
+  end
```
Please add a negative scenario for this spec

BrittanyBlake commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,202 @@
+    end
+  end
+
+  describe "#my_select" do
+    it "Accepts A Proc As Parameter" do
+      expect(
+        [1, 2, 3, 4, 5].my_select(&:even?)
+      ).to(eq([2, 4]))
+    end
+
+    it "Accepts A Proc As Parameter Looping A Range" do
+      expect(
+        (0..10).my_select(&:even?)
+      ).to(eq([0, 2, 4, 6, 8, 10]))
+    end
+  end
```
Please add a negative scenario for this spec

BrittanyBlake commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,202 @@
+      ).to(eq(true))
+    end
+  end
+  describe '#my_count' do
+    it 'Counts elements in array' do
+      expect([1, 2, 4, 2].my_count).to(eq(4))
+    end
+
+    it 'Counts elements matching a pattern' do
+      expect([1, 2, 4, 2].my_count(2)).to(eq(2))
+    end
+
+    it 'Counts elements matching a proc' do
+      expect([1, 2, 4, 2].my_count(&:even?)).to(eq(3))
+    end
+  end
```
Please add a negative scenario.

BrittanyBlake reviewed (changes requested):
### Status: Changes Required ♻️ 
****
### Hi @s0kil  and @KelynPNjeri! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with your project so far! You have been extremely thorough with your test cases and they all are working as they should👏 👏 👏  There is just one issue that needs to be resolved before you can move on: ⬇️ 
****

- [x] Please ensure that all of your specs have negative scenarios that check for the incorrect usage of the methods. In the comments, I have highlighted some areas where they are missing. 

****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

cod3smith committed:
Add negative scenarios for #my_count, #my_none and #my_select.

cod3smith committed:
Add tests for #my_each method.
This adds the required negative scenarios to the test methods.
Tests were added ensuring that all the return values from the my_each enumerable method were correct.

cod3smith committed:
Fix rubocop issues.

cod3smith committed:
Update enumerable test.
Improvement on the pre-existing tests that only tested positive scenarios.
Herein are the negative scenarios for the enumerable projects.

BrittanyBlake reviewed (approved):
### Status: Approved ✅ 
****
### Hi Team! 👋 
### Great job implementing the requested changes!! 🎖️ It's a 👍 from me! Go ahead and merge this pull request and keep up the good work 😄
****
 Good Luck and Happy Coding! 🍀 👩🏾‍💻
